### PR TITLE
DAOS-10397 test: Add ucx provider support to dmg_network_scan.py

### PR DESCRIPTION
**Pull request by phender**
Adding new network_utils.py methods to collect network information for
ofi and ucx andproviding a data structure for storing this information.

Updating the dmg_network_scan.py to make use of these new utility
methods to verify interfaces used with ofi and ucx providers.

Skip-unit-tests: true
Test-tag: dmg,network_scan,basic

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

File: src/tests/ftest/control/dmg_network_scan.py [hunk 1]
```diff
@@ -4,273 +4,72 @@
...
-                cfg_p = "ofi+verbs;ofi_rxm"
-            f_net_devs = [dev for dev in f_net_devs if dev.providers == [cfg_p]]
-
-        return f_net_devs
+        server_provider = self.server_managers[0].get_config_value("provider")
+        sys_info = []
+        for entry in get_network_information(self.hostlist_servers, SUPPORTED_PROVIDERS):
+            if entry.device.startswith("ib") and server_provider in entry.provider:
+                sys_info.append(entry)
+                sys_info[-1].ib_device = None
```
Comment (daltonbohning) [inline src/tests/ftest/control/dmg_network_scan.py:38]: Optional - Personally, I think this is more clear
```suggestion
                entry.ib_device = None
                sys_info.append(entry)
```

File: src/tests/ftest/control/dmg_network_scan.py [hunk 2]
```diff
@@ -4,273 +4,72 @@
...
-        JIRA ID: DAOS-2516
+        """JIRA ID: DAOS-2516
+
         Test Description: Test basic dmg functionality to scan the network
         devices on the system.
-        :avocado: tags=all,small,daily_regression,hw,dmg,network_scan,basic
+
+        :avocado: tags=all,daily_regression
+        :avocado: tags=hw,small
+        :avocado: tags=dmg,network_scan,basic
```
Comment (daltonbohning) [inline src/tests/ftest/control/dmg_network_scan.py:59]: We've been slowly adding the `control` feature

```suggestion
        :avocado: tags=all,daily_regression
        :avocado: tags=hw,small
        :avocado: tags=dmg,control
        :avocado: tags=network_scan,basic
```

File: src/tests/ftest/util/general_utils.py [hunk 3]
```diff
@@ -288,28 +288,107 @@ def run_command(command, timeout=60, verbose=True, raise_exception=True,
...
+def check_task(task, logger=None):
+    """Check the results of the executed task and get the output.
+
+    Args:
+        task (Task): a ClusterShell.Task.Task object for the executed command
+
+    Returns:
+        tuple:
+            bool: if the command returned an 0 exit status on every host
+            str: Summary of the task output
```
Comment (daltonbohning) [inline src/tests/ftest/util/general_utils.py:328]: Looks like it just returns `bool`
`return len(results) == 1 and 0 in results`

File: src/tests/ftest/util/network_utils.py [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+        """Override the default implementation to compare devices."""
+        status = isinstance(other, NetworkDevice)
+        for key in self.__dict__:
+            if not status:
+                break
+            try:
+                status &= getattr(self, key) == getattr(other, key)
+            except AttributeError:
+                status = False
+        return status
```
Comment (daltonbohning) [inline src/tests/ftest/util/network_utils.py:56]: Optional - I realize you only moved this function, but this implementation bothers me. I think just a simple `==` should work here, and is more complete
```suggestion
    def __eq__(self, other):
        """Override the default implementation to compare devices."""
        if not isinstance(other, NetworkDevice):
            return False
        return self.__dict__ == other.__dict__
```

**Review by daltonbohning — CHANGES_REQUESTED** [review]

File: src/tests/ftest/util/general_utils.py [hunk 5]
```diff
@@ -288,28 +288,107 @@ def run_command(command, timeout=60, verbose=True, raise_exception=True,
...
     task = task_self()
     # Enable forwarding of the ssh authentication agent connection
     task.set_info("ssh_options", "-oForwardAgent=yes")
-    kwargs = {"command": command, "nodes": NodeSet.fromlist(hosts)}
+    kwargs = {"command": command, "nodes": hosts}
     if timeout is not None:
         kwargs["timeout"] = timeout
+    if verbose:
+        log = getLogger()
+        log.info("Running on {}: {}".format(hosts, command))
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/general_utils.py:314]: Fixed

File: src/tests/ftest/util/network_utils.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+        list: a list of NetworkDevice objects identifying the network devices on each host
+
+    """
+    network_devices = []
+
+    ofi_info = get_ofi_info(hosts, supported, verbose)
+    ucx_info = get_ucx_info(hosts, supported, verbose)
+
+    interfaces = get_active_network_interfaces(hosts, verbose)
+    for host in hosts:
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/network_utils.py:500]: Fixed

File: src/tests/ftest/control/dmg_network_scan.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/control/dmg_network_scan.py:38]: Done

File: src/tests/ftest/control/dmg_network_scan.py (hunk 2 above)
Comment (phender) [reply src/tests/ftest/control/dmg_network_scan.py:59]: Done

File: src/tests/ftest/util/general_utils.py [hunk 7]
```diff
@@ -288,28 +288,107 @@ def run_command(command, timeout=60, verbose=True, raise_exception=True,
...
+
+    # Display the command output
+    for code in sorted(results):
+        output_data = list(task.iter_buffers(results[code]))
+        if not output_data:
+            output_data = [["<NONE>", results[code]]]
+        for output, o_hosts in output_data:
+            node_set = NodeSet.fromlist(o_hosts)
+            lines = []
+            lines = list(output.splitlines())
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/general_utils.py:351]: Thanks, missed that one.

File: src/tests/ftest/util/network_utils.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/util/network_utils.py:56]: I'm missing an additional edit here, so I do need a bit of extra code, but I can clean it up.

File: src/tests/ftest/util/network_utils.py [hunk 8]
```diff
@@ -45,14 +45,14 @@ def __ne__(self, other):
...
+            return False
         for key in self.__dict__:
-            if not status:
-                break
             try:
-                status &= getattr(self, key) == getattr(other, key)
+                status &= str(getattr(self, key)) == str(getattr(other, key))
             except AttributeError:
-                status = False
+                return False
```
Comment (daltonbohning) [inline src/tests/ftest/util/network_utils.py:55]: Optional - The original implementation had `if not status: break` and then `return status`. Which meant that if at any point `status` was `False`, the function would return `False`. So I think we can remove the `status` variable entirely and just directly return `False`

**Review by daltonbohning — APPROVED** [review]

**Comment by phender** [comment]
Testing the same changes with ucx in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8725/23/

File: src/tests/ftest/util/network_utils.py [hunk 9]
```diff
@@ -509,6 +509,7 @@ def get_network_information(hosts, supported=None, verbose=True):
             for ib_name in data_gather["ib_device"]:
                 for add_on in ([], ["1"]):
                     device = ":".join([ib_name] + add_on)
+                    data_gather["provider"].update(get_interface_providers(device, ofi_info))
                     data_gather["provider"].update(get_interface_providers(device, ucx_info))
```
Comment (daltonbohning) [inline src/tests/ftest/util/network_utils.py:512]: Doesn't the second update override the first?

Comment (phender) [reply src/tests/ftest/util/network_utils.py:512]: No, it appends because the providers are different

**Comment by phender** [comment]
Testing passed with verbs (https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-9025/9/artifact/Functional%20Hardware%20Small/control/dmg_network_scan.py):
```
2022-05-20 19:36:08,390 dmg_network_scan L0070 INFO | ----------------------------------------------------------------------------------------------------
2022-05-20 19:36:08,390 dmg_network_scan L0071 INFO | SYS INFO
2022-05-20 19:36:08,390 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib0, ib_device=None, port=1, provider=ofi+verbs;ofi_rxm, numa=0)
2022-05-20 19:36:08,391 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib1, ib_device=None, port=1, provider=ofi+verbs;ofi_rxm, numa=1)
2022-05-20 19:36:08,391 dmg_network_scan L0070 INFO | ----------------------------------------------------------------------------------------------------
2022-05-20 19:36:08,391 dmg_network_scan L0071 INFO | DMG INFO
2022-05-20 19:36:08,391 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib0, ib_device=None, port=1, provider=ofi+verbs;ofi_rxm, numa=0)
2022-05-20 19:36:08,391 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib1, ib_device=None, port=1, provider=ofi+verbs;ofi_rxm, numa=1)
```
and ucx (https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-8725/23/artifact/Functional%20Hardware%20Small/control/dmg_network_scan.py):
```
2022-05-20 18:53:12,963 dmg_network_scan L0070 INFO | ----------------------------------------------------------------------------------------------------
2022-05-20 18:53:12,963 dmg_network_scan L0071 INFO | SYS INFO
2022-05-20 18:53:12,963 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib0, ib_device=None, port=1, provider=ucx+dc_x, numa=0)
2022-05-20 18:53:12,963 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib1, ib_device=None, port=1, provider=ucx+dc_x, numa=1)
2022-05-20 18:53:12,963 dmg_network_scan L0070 INFO | ----------------------------------------------------------------------------------------------------
2022-05-20 18:53:12,963 dmg_network_scan L0071 INFO | DMG INFO
2022-05-20 18:53:12,963 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib0, ib_device=None, port=1, provider=ucx+dc_x, numa=0)
2022-05-20 18:53:12,963 dmg_network_scan L0073 INFO |   NetworkDevice(host=wolf-110, device=ib1, ib_device=None, port=1, provider=ucx+dc_x, numa=1)
2022-05-20 18:53:12,964 dmg_network_scan L0074 INFO | ----------------------------------------------------------------------------------------------------
```

File: src/tests/ftest/util/network_utils.py (hunk 9 above)
Comment (daltonbohning) [reply src/tests/ftest/util/network_utils.py:512]: Ahh, `get_interface_providers` returns a `dict`! :)

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**